server/balancer.go

```diff
@@ -0,0 +1,316 @@
+package server
+
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/kvproto/pkg/raftpb"
+)
+
+// Operator is the interface to do some operations.
+type Operator interface {
+	// Do does the operator, if finished then return true.
+	Do(req *pdpb.RegionHeartbeatRequest) (bool, *pdpb.RegionHeartbeatResponse)
+	// Check checks whether the operator has been finished.
+	Check(req *pdpb.RegionHeartbeatRequest) bool
+}
+
+func containPeer(region *metapb.Region, peer *metapb.Peer) bool {
+	for _, p := range region.GetPeers() {
+		if p.GetId() == peer.GetId() {
+			return true
+		}
+	}
+
+	return false
+}
+
+func leaderPeer(region *metapb.Region, storeID uint64) *metapb.Peer {
+	for _, peer := range region.GetPeers() {
+		if peer.GetStoreId() == storeID {
+			return peer
+		}
+	}
+
+	return nil
+}
+
+// BalanceOperator is used to do region balance.
+type BalanceOperator struct {
+	index int
+	ops   []Operator
+}
+
+func newBalanceOperator(oldLeader, newLeader, newPeer *metapb.Peer) *BalanceOperator {
```

Comment:
This seems for capacity operator only.
As a common abstract, maybe you should create it with Operator directly, like 
```
func newBalanceOperator(ops ...Operator) *BalanceOperator
```

And you should combine the operators outer for special strategy.